odoo/addons/base/models/res_lang.py

```diff
@@ -162,6 +174,185 @@ def fix_datetime_format(format):
+
+                    # TODO different separators for windows in tar?
+                    addons = filename.split('/')[0]
+                    if addons not in module_list:
+                        # ensure a pack does not inject translations from another module
+                        _logger.debug("Skip translations for unexpected module %s", addons)
+                        continue
+
+                    po_lang = filename.split('/')[-1][:-3]
+                    if base_lang_code and po_lang == base_lang_code:
+                        _logger.debug("Extracting base translation file %s to %s", filename, tmp)
+                        downloaded_addons[base_lang_code].add(addons)
+                        tar_content.extract(filename, path=tmp)
+                        src_file = os.path.join(tmp, filename)
+                        with open(src_file, 'rb') as po_file:
+                            tools.trans_load_data(self._cr, po_file, 'po', base_lang_code, module_name=addons, verbose=False)
+
+                    elif po_lang == iso_code:
+                        _logger.debug("Extracting translation file %s to %s", filename, tmp)
+                        downloaded_addons[iso_code].add(addons)
+                        tar_content.extract(filename, path=tmp)
+                        src_file = os.path.join(tmp, filename)
+                        with open(src_file, 'rb') as po_file:
+                            tools.trans_load_data(self._cr, po_file, 'po', iso_code, module_name=addons, verbose=False)
+                    else:
+                        _logger.debug("Skip translations for unexpected language %s", po_lang)
+                        continue
+
+
+        return downloaded_addons
+
+    def _download_translation_files(self):
+        """ Download the translation files of all modules from the i18n servers """
+        mods = self.env['ir.module.module'].search_read([('state', '!=', 'uninstallable'),],
+                                                        fields=['name', 'i18n_location'])
+        # consider language packs are the most up to date, store the list of downloaded
+        # module per language to check the filesystem only if was not found online
+        processed = defaultdict(set)
+
+        # [{'id': 1, 'name': 'base', 'i18n_location': 'https://...'},...] -> {'https://...': ['base',...],...}
+        urls = defaultdict(list)
+        for module_info in mods:
+            urls[module_info['i18n_location']].append(module_info['name'])
+
+        # ['fr_BE', 'fr_FR', 'nl_BE'] -> {'fr': ['fr_FR', 'fr_BE'], 'nl': ['nl_BE']}
+        # will download once and process twice for self = [fr, fr_BE]
+        langs = itertools.groupby(self.mapped('code'), key=lambda l: l.split('_')[0])
+        for short_code, full_codes in langs:
+            for url in urls:
+                full_url = self._get_i18n_url(url, short_code)
+                try:
+                    stream = requests.get(full_url, timeout=5)
```

Comment:
Might be a good idea to use a conditional request here e.g. store the last-modified and etag when downloading a file and feed it back on the next download. 

& yesterday wondered why not download to `data_dir` the way it's done for appstore, then load from there when e.g. installing new modules? That way we ought be able to share the bundle between databases / instances, and can more easily manage allowing or forbidding db owners access to this feature or somesuch e.g. `$data_dir/translations/<version>/<lang>/...`, possibly adding some sort of descriptor file storing etag & last-modified for conditional request-ing?